File: two-pointers/0234-palindrome-linked-list/palindrome-linked-list-solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        
        /*
        first we find the middle of the list 
        then ignore the mid term if odd numbers are there in the list 
        split the list into two 
        reverse the second list 
        compare the two lists now
        */
        ListNode* slow = head;
        ListNode* fast = head;
        stack<int> st;
        while(fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        ListNode* firstHalf = head;
        ListNode* secondHalf;
        
        if(fast)
            secondHalf = slow->next;
        else
            secondHalf = slow;
        
        while(firstHalf != slow)
        {
            st.push(firstHalf->val);
            firstHalf = firstHalf->next;
        }
        
        while(secondHalf)
        {
            if(st.empty() || st.top() != secondHalf->val)
                return false;
            
            st.pop();
            secondHalf = secondHalf->next;
        }
        
        return true;
    }
};

File: two-pointers/0234-palindrome-linked-list/palindrome-linked-list-solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        fast = reverseListNode(slow);
        slow = head;

        while (slow != null && fast != null) {
            if (slow.val != fast.val) return false;
            slow = slow.next;
            fast = fast.next;
        }
        return true;
    }

    public static ListNode reverseListNode (ListNode head) {
        ListNode curr = head;
        ListNode prev = null;
        ListNode next = null;

        // traverse all the nodes of LinkedList
        while (curr != null) {

            // store next
            next = curr.next;

            // reverse current nodes next pointer
            curr.next = prev;

            // move pointers one position ahead
            prev = curr;
            curr = next;
        }

        return prev;
    }
}
